rfcs/004 - Continuous Two-Way Synchronization RFC.md

```diff
@@ -0,0 +1,163 @@
+2. We get two-way automatic mode by default so most of the examples only need normal-looking SQL statements, making onboarding easier.
+3. We get both one-way automatic modes not only for the whole database, but even at table and row granularities, giving us advantages over existing IaC tools (continuously fix failures by restoring infrastructure for you, but also make sure you don't delete resources not owned by your team by blocking mutation of them).
+4. The database, in an eventually-consistent way, always reflects the state of your cloud, so querying is straightforward.
+5. Two-way mode auto-importing resources created by "other" things should make auto-generating limited modules possible (create an entity from a `List-` AWS API call, then turn every other AWS API into an RPC call from the database, allowing imperative Create/Update/Delete and a read-only table to SELECT results from that is re-synced with the cloud every so often), which will let us more rapidly hit 100% cloud coverage (though with a second-tier of quality), which can help us with larger startups and enterprises that need full coverage more than they need cleaner, mutable tables with explicit dependency foreign keys.
+
+That last piece is particularly interesting because it's a way to help us accelerate our development velocity and potentially turn our current "low-level modules" into "mid-level modules" that could be pure SQL modules depending on these RPC functions, and the "high-level modules" could decide to depend on either of them, depending on what is easier.
+
+### Alternatives Considered
+
+#### Just don't implement any `code-` modules
+
+Basically, we can ~~stick our heads in the sand~~ decide to enforce management of all cloud resources in a way that these changes are represented in our database, but this makes our ability to co-exist with other IaC tools impossible, as well as integrating with any sort of CI system (without basically re-implementing our own CI system we then demand users use). And therefore this approach has been rejected.
+
+#### Implement `iasql_commit` as replacement for `iasql_apply`/`iasql_sync` only
+
+Stick to manual mode and just replace `apply` and `sync` with `iasql_commit`. This allows us to implement the `code-` modules and requires less work, but keeps the harder to explain syntax in the forefront, rather than being a power user tool. It also precludes our ability to support Enforce/Track behavior described above, which narrows our set of advantages.
+
+This may still be a point along-the-way in the conversion to the Full Two-Way Mode, though, but has been rejected as the stopping point.
+
+#### Implement Two-Way Mode with the transaction-like syntax only
+
+Basically just the first half of the proposal, stopping before the Enforce/Track section. This is probably also a point along-the-way, but has been rejected as the stopping point because the remaining features are simpler to implement and provide significant functionality.
```

Comment:
I think this is outdated to reflect the new implementation proposal via the start/end timestamp check and we should probably call the old version out too?

Reply:
I will double-check that these are referring to the same sections as they used to, but I don't follow "the old version" here?